src/language/class-modifiers.md

```diff
@@ -0,0 +1,258 @@
+
+Only the `base` modifier can appear before a mixin declaration. The modifiers do
+not apply to other declarations like `enum`, `typedef`, or `extension`.
+
+When deciding whether to use class modifiers, consider the intended uses of the
+class, and what behaviors the class needs to be able to rely on.
+
+## No modifier
+
+Use a `class` or `mixin` declaration without a modifier to allow unrestricted
+permission to create subtype relationships from any library.
+By default, you can:
+
+- [Construct][] new instances of a class.
+- [Extend][] a class to create a new subtype.
+- [Implement][] a class or mixin's interface.
+- [Mix in][mixin] a mixin or mixin class.
+
+## `abstract`
+
+Use the `abstract` modifier to define a class that can’t be constructed from any
+library (it's own, as well as outside libraries).
+Abstract classes are useful for defining interfaces, often with some
+implementation. Abstract classes often have [abstract methods][].
+
+```dart
+// Library a.dart
+abstract class Vehicle { 
+  void moveForward();     // Abstract method.
+}
+```
+
+```dart
+// Library b.dart
+import 'a.dart';
+
+var myCar = Vehicle();     // Error: Cannot be constructed
+
+class Car extends          // Can be extended
+  Vehicle {
+    int passengers;
+    ...
+}
+
+class MockVehicle          // Can be implemented
+  implements Vehicle {
+    @override
+    void moveForward(int meters) { ... }
+}
+```
+
+If you want your abstract class to appear to be instantiable,
+define a [factory constructor][].
+
+## `base`
+
+Use the `base` modifier to enforce inheritance of a class or mixin's implementation.
+The `base` modifier disallows implementation outside of the current library, which
+guarantees:
+
+- The base class constructor is called whenever an instance of a subtype of the
+class is created.
+- That all implemented private members exist in subtypes.
+- Adding a new implemented member to a `base` class will not break subtypes (except
+  when they already declare a member with the same name and an incompatible
+  signature), since all subtypes will inherit the new member.
+
+The `base` modifier must be applied transitively to all the subtypes of a class;
+any class which implements or extends a base class must be marked `base`, `final`,
+or `sealed`. 
+
+```dart
+// Library a.dart
+base class Vehicle {
+  void moveForward(int meters) { ... }
+}
+```
+
+```dart
+// Library b.dart
+import 'a.dart';
+
+var myCar = Vehicle();          // Can be constructed
+
+base class Car extends          // Can be extended
+  Vehicle {
+    int passengers;
+    ...
+}
+
+base class MockVehicle          // ERROR: Cannot be implemented
+  implements Vehicle {
+    @override
+    void moveForward ...
+}
+```
+
+## `interface`
+
+Use the `interface` modifier to define an interface. Interface classes can be
+implemented outside of the current library, but not extended. This guarantees:
+
+- All members of the class must have an implementation.
```

Comment:
that's actually quite surprising, if you're coming from other languages with an `interface` keyword. (Most other languages' `interface` keyword means what we mean by `abstract interface class`.) Might be worth acknowledging this weirdness.

Reply:
There are some sources of ambiguity here. 'All members of the class' would generate the question 'Which class?'.

You can certainly have an `abstract interface class A ...` where some members do not have an implementation. With a plain `interface class C` it is required that `C` is a concrete class (simply because we didn't say `abstract`), and in that case all its members must have an implementation.

But this is all about `abstract` or not, which means that it is confusing to mention it as if it had something to do with `interface`.

I guess the point is that we're allowed to have implementations of members in an `interface class` and we're still guaranteed (assuming no reopening) that nobody can inherit those member implementations. This means that we have more freedom to change that implementation.

In other languages you'd probably write an `interface I` and then `private class C implements I`, and that's how you would ensure that nobody can inherit the instance member implementations. With a Dart `interface class` we can do both in the same declaration. This may be seen as ugly or not, but you will at least be able to avoid duplicating a bunch of member signatures (that is: the parameter types, return type and name, for each member).